### PR TITLE
KAFKA-6435: KIP-623 Add internal topics option to streamResetter

### PR DESCRIPTION
**Pull request by JoelWee**
This allows users to specify which internal-topics the tool will attempt to delete instead of inferring it.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by abbccdda** [comment]
@JoelWee Could we rebase this work?

**Comment by JoelWee** [comment]
@abbccdda rebased :)

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 1]
```diff
@@ -646,22 +655,68 @@ private boolean isIntermediateTopic(final String topic) {
...
+        System.out.println("Following inferred internal/auto-created topics will be deleted for application " + options.valueOf(applicationIdOption));
+        for (final String topic : topicsToDelete) {
+            System.out.println("Topic: " + topic);
+        }
+
         if (!dryRun) {
             doDelete(topicsToDelete, adminClient);
         }
         System.out.println("Done.");
+        return EXIT_CODE_SUCCESS;
```
Comment (abbccdda) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: We could refactor out a helper function here.

File: docs/streams/developer-guide/app-reset-tool.html [hunk 2]
```diff
@@ -77,6 +77,7 @@
         <div class="section" id="step-1-run-the-application-reset-tool">
             <h2>Step 1: Run the application reset tool<a class="headerlink" href="#step-1-run-the-application-reset-tool" title="Permalink to this headline"></a></h2>
             <p>Invoke the application reset tool from the command line</p>
+            <p>Warning! This tool makes irreversible changes to your application. It is strongly recommended that you run this once with --dry-run to preview your changes before making them.</p>
```
Comment (abbccdda) [inline docs/streams/developer-guide/app-reset-tool.html]: for html changes, it is recommended to include a screenshot of the built website. Please refer to https://cwiki.apache.org/confluence/display/KAFKA/Setup+Kafka+Website+on+Local+Apache+Server

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java [hunk 3]
```diff
@@ -262,6 +263,52 @@ public void shouldNotAllowToResetWhenIntermediateTopicAbsent() throws Exception
...
+        // RESET
+        streams.cleanUp();
+
+        final List<String> internalTopics = cluster.getAllTopicsInCluster().stream()
+                .filter(topic -> topic.startsWith(appID + "-"))
+                .collect(Collectors.toList());
+        final boolean cleanResult = tryCleanGlobal(!useRepartitioned,
+                "--internal-topics",
+                String.join(",", internalTopics.subList(1, internalTopics.size())) + "," + OUTPUT_TOPIC);
+        Assert.assertEquals(false, cleanResult); // Reset will give error code since output topic is not a valid internal topic
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: could use `assertFalse`

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java [hunk 4]
```diff
@@ -262,6 +263,52 @@ public void shouldNotAllowToResetWhenIntermediateTopicAbsent() throws Exception
...
+        streams.close();
+        waitForEmptyConsumerGroup(adminClient, appID, TIMEOUT_MULTIPLIER * STREAMS_CONSUMER_TIMEOUT);
+
+        // RESET
+        streams.cleanUp();
+
+        final List<String> internalTopics = cluster.getAllTopicsInCluster().stream()
+                .filter(topic -> topic.startsWith(appID + "-"))
+                .collect(Collectors.toList());
+        final boolean cleanResult = tryCleanGlobal(!useRepartitioned,
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: Could just use `false`

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 5]
```diff
@@ -646,22 +655,68 @@ private boolean isIntermediateTopic(final String topic) {
...
-                } else {
-                    System.out.println("Topic: " + listing);
-                }
+            if (isInferredInternalTopic(listing)) {
+                topicsToDelete.add(listing);
             }
         }
+
+        System.out.println("Following inferred internal/auto-created topics will be deleted for application " + options.valueOf(applicationIdOption));
+        for (final String topic : topicsToDelete) {
```
Comment (abbccdda) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Could we print all topics in one line, instead of multiple?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 6]
```diff
@@ -646,22 +655,68 @@ private boolean isIntermediateTopic(final String topic) {
...
+        System.out.println("Deleting specified internal/auto-created topics " + internalTopics);
+        for (final String topic : internalTopics) {
+            if (allTopics.contains(topic) && isInferredInternalTopic(topic)) {
+                topicsToDelete.add(topic);
+            } else {
+                notFoundInternalTopics.add(topic);
+            }
+        }
+
+        if (!notFoundInternalTopics.isEmpty()) {
```
Comment (abbccdda) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Should we throw exception here to stop the run or just skip the not found topics?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 7]
```diff
@@ -167,7 +171,7 @@ public int run(final String[] args,
             final HashMap<Object, Object> consumerConfig = new HashMap<>(config);
             consumerConfig.putAll(properties);
             exitCode = maybeResetInputAndSeekToEndIntermediateTopicOffsets(consumerConfig, dryRun);
-            maybeDeleteInternalTopics(adminClient, dryRun);
+            exitCode |= maybeDeleteInternalTopics(adminClient, dryRun);
```
Comment (abbccdda) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:173]: should we do check both exit codes and decide whether to return 0 or 1?

**Review by abbccdda — COMMENTED** [review]
Thanks for the PR, got a couple of comments.

File: docs/streams/developer-guide/app-reset-tool.html (hunk 2 above)
Comment (JoelWee) [reply docs/streams/developer-guide/app-reset-tool.html]: 👍  here's a screenshot for this portion:

Screenshot 2020-08-13 at 2 19 01 PM

File: docs/streams/developer-guide/app-reset-tool.html [hunk 8]
```diff
@@ -106,6 +107,11 @@ <h2>Step 1: Run the application reset tool<a class="headerlink" href="#step-1-ru
                                         topics <span class="o">(</span>topics used in the through<span class="o">()</span>
                                         method<span class="o">)</span>. For these topics, the tool
                                         will skip to the end.
+--internal-topics &lt;String: list&gt;      Comma-separated list of internal topics
+                                        to delete. Must be a subset of the
+                                        internal topics marked for deletion by
+                                        the default behaviour (do a dry-run without
+                                        this option to view these topics).
```
Comment (JoelWee) [inline docs/streams/developer-guide/app-reset-tool.html:113]: screenshot of this portion:

Screenshot 2020-08-13 at 2 19 35 PM

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 7 above)
Comment (JoelWee) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:173]: Yep. Currently it returns 1 if either exitCode is 1, and 0 otherwise. Or should we do something else?

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 6 above)
Comment (JoelWee) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: I would prefer to throw an exception, since it's likely that the user has made a mistake. I've implemented this now

**Comment by JoelWee** [comment]
Hi @abbccdda, would you mind taking another look? :)

**Comment by ableegoldman** [comment]
Hey @abbccdda , can you take another look at this? @JoelWee are you still interested in getting this KIP done? 

Maybe one of @cadonna @mjsax @lct45 @wcarlson5 can help with the review if so

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 9]
```diff
@@ -248,7 +257,7 @@ private void parseArguments(final String[] args) {
             .ofType(String.class)
             .describedAs("file name");
         forceOption = optionParser.accepts("force", "Force the removal of members of the consumer group (intended to remove stopped members if a long session timeout was used). " +
-                "Make sure to shut down all stream applications when this option is specified to avoid unexpected rebalances.");
```
Comment (wcarlson5) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: unnecessary whitespace change

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 10]
```diff
@@ -225,6 +229,11 @@ private void parseArguments(final String[] args) {
             .ofType(String.class)
             .withValuesSeparatedBy(',')
             .describedAs("list");
+        internalTopicsOption = optionParser.accepts("internal-topics", "Comma-separated list of internal topics to delete. Must be a subset of the internal topics marked for deletion by the default behaviour (do a dry-run without this option to view these topics).")
```
Comment (wcarlson5) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: line is too long

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 11]
```diff
@@ -642,22 +651,38 @@ private boolean isIntermediateTopic(final String topic) {
...
+                .filter(this::isInferredInternalTopic)
+                .collect(Collectors.toList());
+        final List<String> specifiedInternalTopics = options.valuesOf(internalTopicsOption);
+        final List<String> topicsToDelete;
+
+        if (!specifiedInternalTopics.isEmpty()) {
+            final List<String> notFoundInternalTopics = specifiedInternalTopics.stream()
+                    .filter(topic -> !inferredInternalTopics.contains(topic))
+                    .collect(Collectors.toList());
+            if (!notFoundInternalTopics.isEmpty()) {
```
Comment (wcarlson5) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: something like `inferredInternalTopics.containsAll(specifiedInternalTopics)` might be easier to understand here

**Review by wcarlson5 — APPROVED** [review]
Thanks for the KIP @JoelWee !

I had a few nits, but overall very nice job. Can you rebase and resolve the conflicts as well?

@ableegoldman if you agree can you give this a merge?

**Comment by JoelWee** [comment]
Thanks @ableegoldman and @wcarlson5! Will try to get this done over the next week or so and ping you when ready :)

**Comment by JoelWee** [comment]
Hi @ableegoldman, have rebased now :). The failing tests are RaftClusterTests which don't seem related to the changes here?

File: docs/streams/developer-guide/app-reset-tool.html [hunk 12]
```diff
@@ -78,6 +78,9 @@
             <h2>Step 1: Run the application reset tool<a class="headerlink" href="#step-1-run-the-application-reset-tool" title="Permalink to this headline"></a></h2>
             <p>Invoke the application reset tool from the command line</p>
             <div class="highlight-bash"><div class="highlight"><pre><span></span><code>&lt;path-to-kafka&gt;/bin/kafka-streams-application-reset</code></pre></div>
+            <p>Warning! This tool makes irreversible changes to your application. It is strongly recommended that you run this once with <code class="docutils literal"><span class="pre">--dry-run</span></code> to preview your changes before making them.</p>
+            <div class="highlight-bash"><div class="highlight"><pre><span></span>&lt;path-to-kafka&gt;/bin/kafka-streams-application-reset
```
Comment (ableegoldman) [inline docs/streams/developer-guide/app-reset-tool.html]: You don't need to repeat this line, it's just printing an example of running the app reset tool for the line above (`Invoke the application reset tool from the command line`)

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java [hunk 13]
```diff
@@ -205,6 +206,34 @@ private void add10InputElements() {
...
+        IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(resultConsumerConfig, OUTPUT_TOPIC, 10);
+
+        streams.close();
+        waitForEmptyConsumerGroup(adminClient, appID, TIMEOUT_MULTIPLIER * STREAMS_CONSUMER_TIMEOUT);
+
+        // RESET
+        streams.cleanUp();
+
+        final List<String> internalTopics = cluster.getAllTopicsInCluster().stream()
+                .filter(topic -> topic.startsWith(appID + "-"))
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: nit: we should also filter for internal topics specifically, like what's done in `StreamsResetter#matchesInternalTopicFormat`. Actually you can probably just invoke that method directly for the filter here (it can be made static if necessary)

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java [hunk 14]
```diff
@@ -151,6 +151,22 @@ public void shouldNotAllowToResetWhenIntermediateTopicAbsent() {
         Assert.assertEquals(1, exitCode);
     }
 
+    @Test
+    public void shouldNotAllowToResetWhenSpecifiedInternalTopicAbsent() {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: Just wondering, why put this test here instead of in `AbstractResetIntegrationTest`?

Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: Also, can we add a test like this but for the case where the topic does exist but just isn't a subset of inferred internal topics?

**Review by ableegoldman — COMMENTED** [review]
Looks good for the most part, just a few suggestions. Also, this is unrelated to your PR but I just happened to notice it: there's a comment in `AbstractResetIntegrationTest.tryCleanGlobal` that starts with `//leaving --zookeeper arg here...`, can you delete it? This config has been deprecated for a while and I removed it in a recent PR but clearly missed this comment 😅

File: docs/streams/developer-guide/app-reset-tool.html (hunk 12 above)
Comment (JoelWee) [reply docs/streams/developer-guide/app-reset-tool.html]: Thanks! Have updated this now to look like:

Screenshot 2021-04-27 at 7 54 48 PM

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java (hunk 13 above)
Comment (JoelWee) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: done now :)

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java (hunk 14 above)
Comment (JoelWee) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: It seems more natural to group it together with the other `shouldNotAllowToResetWhen...` tests. E.g. shouldNotAllowToResetWhenIntermediateTopicAbsent, shouldNotAllowToResetWhenInputTopicAbsent, etc.

Happy to shift it over to AbstractResetIntegrationTest.java

Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: Hah, maybe I should have asked why are all of those tests not also in `AbstractResetIntegrationTest`. Seems like almost everything that applies here would also be good to test in the SSL version of the test (which AFAICT is the only other one to extend the AbstractResetIntegrationTest).

But I'm ok with leaving it as is, and maybe we can just look into this as followup work unless there is a good reason for them to be where they are (which I can't think of but my imagination is not endless)

**Comment by ableegoldman** [comment]
Oof, there are a LOT of flaky test failures in this build. They're all unrelated to this PR, mostly in Connect and the RaftClusterTest, so I'll go ahead and merge, but yikes.

**Review by ableegoldman — APPROVED** [review]
LGTM, happy to get this one in at last!

**Comment by JoelWee** [comment]
Thanks @ableegoldman! :)
